packages/astro/src/vite-plugin-astro-server/app.ts

```diff
@@ -0,0 +1,531 @@
+				return error;
+			},
+		});
+	}
+
+	async handleRoute({
+		matchedRoute,
+		incomingRequest,
+		incomingResponse,
+		body,
+		url,
+		pathname,
+	}: HandleRoute): Promise<void> {
+		const timeStart = performance.now();
+		const { config, loader, logger } = this.pipeline;
+
+		if (!matchedRoute) {
+			// This should never happen, because ensure404Route will add a 404 route if none exists.
+			throw new Error('No route matched, and default 404 route was not found.');
+		}
+
+		let request: Request;
+		let renderContext: RenderContext;
+		let route: RouteData = matchedRoute.route;
+		const componentInstance = await this.pipeline.preload(route, matchedRoute.filePath);
+		const actions = await loadActions(loader);
+		this.pipeline.setActions(actions);
+		const middleware = (await loadMiddleware(loader)).onRequest;
+		// This is required for adapters to set locals in dev mode. They use a dev server middleware to inject locals to the `http.IncomingRequest` object.
+		const locals = Reflect.get(incomingRequest, clientLocalsSymbol);
+
+		// Allows adapters to pass in locals in dev mode.
+		request = createRequest({
+			url,
+			headers: incomingRequest.headers,
+			method: incomingRequest.method,
+			body,
+			logger,
+			isPrerendered: route.prerender,
+			routePattern: route.component,
+		});
+
+		// Set user specified headers to response object.
+		for (const [name, value] of Object.entries(config.server.headers ?? {})) {
+			if (value) incomingResponse.setHeader(name, value);
+		}
+
+		renderContext = await RenderContext.create({
+			locals,
+			pipeline: this.pipeline,
+			pathname,
+			middleware: isDefaultPrerendered404(matchedRoute.route) ? undefined : middleware,
+			request,
+			routeData: route,
+			clientAddress: incomingRequest.socket.remoteAddress,
+			actions,
+		});
+
+		let response;
+		let statusCode = 200;
+		let isReroute = false;
+		let isRewrite = false;
+
+		try {
+			response = await renderContext.render(componentInstance);
+			isReroute = response.headers.has(REROUTE_DIRECTIVE_HEADER);
+			isRewrite = response.headers.has(REWRITE_DIRECTIVE_HEADER_KEY);
+			const statusCodedMatched = getStatusByMatchedRoute(route);
+			statusCode = isRewrite
+				? // Ignore `matchedRoute` status for rewrites
+					response.status
+				: // Our internal noop middleware sets a particular header. If the header isn't present, it means that the user have
+					// their own middleware, so we need to return what the user returns.
+					!response.headers.has(NOOP_MIDDLEWARE_HEADER) && !isReroute
+					? response.status
+					: (statusCodedMatched ?? response.status);
+		} catch (err: any) {
+			response = await this.renderError(request, {
+				skipMiddleware: false,
+				locals,
+				status: 500,
+				prerenderedErrorPageFetch: fetch,
+				clientAddress: incomingRequest.socket.remoteAddress,
+				error: err,
+			});
+			statusCode = 500;
+		} finally {
+			this.currentRenderContext?.session?.[PERSIST_SYMBOL]();
```

Comment:
The session tests are failing because `this.currentRenderContext` is undefined here. Changing it to `renderContext?.session` makes the tests pass, but I don't know about any other implications of that.